RavenDB-14127 properly notify listeners about database record apply

karmeli87 opened the pull request:
- Don't use `ContinueWith`, but wait for the the database to be loaded before notifying success.
- Rethrow exceptions, so they will be properly notified.

karmeli87 committed:
RavenDB-14127 handle synchronously database record changes

karmeli87 committed:
RavenDB-14127 ensure that `_rachisLogIndexNotifications.AddTask` happened under lock

`OnDispose` ins't under lock, so it might happened that we are awaiting on an index that isn't in the dictonary, and when the index not in the dictionary we assume the task is completed.

karmeli87 committed:
RavenDB-14127 Fix test and show the full stack trace exception

ayende commented on src/Raven.Server/Documents/DatabasesLandlord.cs:
```diff
@@ -108,54 +108,28 @@ private void HandleClusterDatabaseChanged(string databaseName, long index, strin
                 if (task.IsCanceled || task.IsFaulted)
                     return;
 
+                var database = task.Result;
```
This is probably a bad idea. What if we have a long database load time?
You cannot force a wait here.

karmeli87 replied:
`HandleClusterDatabaseChanged` method is executed asynchronously, so it is not blocking any other operation, but it might eat-up threads.. would `AsyncHelpers.RunSync(() => task)` help here?

ayende replied:
No, would also hold us the thread.
Instead, change the execute to be async, so we can await there.

karmeli87 committed:
RavenDB-14127 wait for the database to load in an async manner

1. This change has required us to move from the `ReaderWriterLockSlim`, because it is thread-affine to the `AsyncReaderWriterLock`.
2. Since we changed the return type from `void` to `Task` we also had to move from the `EventHandler` to a more "manual" event notification.

karmeli87 committed:
RavenDB-14127 revert the logic of `ErrorOnDatabaseDoesNotExists`

I guess we want to ignore `DatabaseDoesNotExists` upon deletion, but not when we update the database record

karmeli87 committed:
RavenDB-14127 don't failover upon `DatabaseLoadFailureException` during adding a new node

ayende commented on src/Raven.Server/Documents/DatabasesLandlord.cs:
```diff
@@ -52,140 +53,113 @@ public DatabasesLandlord(ServerStore serverStore)
+                    if (task.IsCanceled || task.IsFaulted)
+                        return;
+
+                    var database = await task;
```
If this fails, the user is going to get an error, but does it impact anything else?

karmeli87 replied:
It might only impact on the server start-up, I guess it is better to create a task for each database.

ayende commented on src/Raven.Server/Documents/DatabasesLandlord.cs:
```diff
@@ -565,8 +490,22 @@ public Task<DocumentDatabase> TryGetOrCreateResourceStore(StringSegment database
+                var awaiter = _disposing.ReaderLockAsync(cts.Token).GetAwaiter();
+                if (awaiter.IsCompleted == false)
+                {
+                    cts.Cancel();
```
I don't like this.
what happened if between the `IsCompleted` check and the `Cancel()`, you got the lock? Then you have a hanging lock here.

karmeli87 replied:
done

ayende commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -979,29 +1028,29 @@ private void RemoveNodeFromCluster(TransactionOperationContext context, Blittabl
+                    foreach (var task in tasks)
                     {
-                        TaskExecutor.Execute(_ =>
+                        Task.Run(async () =>
```
This is going to block until the task is done running?

karmeli87 replied:
No, it should execute the action inside the `Task.Run` asynchronously.

ayende commented on src/Raven.Server/Web/System/AdminDatabasesHandler.cs:
```diff
@@ -190,7 +190,16 @@ public async Task AddDatabaseNode()
+                {
+                    // the node was added successfully, but failed to start 
+                    // in this case we don't want the request executor of the client to fail-over (so we wouldn't create an additional node)
+                    throw new InvalidOperationException(e.Message);
```
```suggestion
                    throw new InvalidOperationException(e.Message, e);
```

karmeli87 replied:
done

karmeli87 committed:
RavenDB-14127 address PR comments
